public/app/features/scenes/variables/sets/VariableUpdateManager.ts

```diff
@@ -0,0 +1,154 @@
+   */
+  private handleNewVariableValueOptions(variable: SceneVariable, options: VariableValueOption[]) {
+    this.updateVariableStateGivenNewOptions(variable, options);
+
+    const update = this.updating.get(variable.state.name);
+    update?.subscription.unsubscribe();
+
+    this.updating.delete(variable.state.name);
+    this.variablesToUpdate.delete(variable.state.name);
+    this.updateNextBatch();
+  }
+
+  private handleVariableError(variable: SceneVariable, err: Error) {
+    variable.setState({ state: LoadingState.Error });
+    // todo handle properly
+  }
+
+  /**
+   * Check if current value is valid given new options. If not update the value.
+   * TODO: Handle multi valued variables
+   */
+  private updateVariableStateGivenNewOptions(variable: SceneVariable, options: VariableValueOption[]) {
+    if (options.length === 0) {
+      // TODO handle the no value state
+      variable.setState({ value: '?', state: LoadingState.Done });
+      return;
+    }
+
+    const foundCurrent = options.find((x) => x.value === variable.state.value);
+    if (!foundCurrent) {
+      // Current value is not valid. Set to first of the available options
+      variable.setState({ value: options[0].value, text: options[0].label, state: LoadingState.Done });
+    } else {
+      // current value is still ok
+      variable.setState({ state: LoadingState.Done });
+    }
+  }
+
+  /**
+   * Checks if the variable has any dependencies that is currently in variablesToUpdate
+   */
+  private hasDependendencyThatNeedsUpdating(variable: SceneVariable) {
+    const dependencies = this.dependencies.get(variable.state.name);
+
+    if (dependencies) {
+      for (const dep of dependencies) {
+        for (const otherVariable of this.variablesToUpdate.values()) {
+          if (otherVariable.state.name === dep) {
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Extract dependencies from all variables and add those that needs update to the variablesToUpdate map
+   * Then it will start the update process.
+   */
+  updateAll() {
+    for (const variable of this.sceneContext.state.variables) {
+      this.variablesToUpdate.set(variable.state.name, variable);
+
+      if (variable.getDependencies) {
+        this.dependencies.set(variable.state.name, variable.getDependencies());
+      }
+    }
+
+    this.updateNextBatch();
+  }
+
+  /**
+   * Updates dependencies in case they changed
+   * */
+  variableStateChanged(variable: SceneVariable) {
+    if (variable.getDependencies) {
+      this.dependencies.set(variable.state.name, variable.getDependencies());
+    }
+  }
+
+  /**
+   * This will trigger an update of all variables that depend on it.
+   * */
+  variableValueChanged(variable: SceneVariable) {
+    // Ignore this change if it is currently updating
+    if (this.updating.has(variable.state.name)) {
+      return;
+    }
```

Comment:
Wonder why this is needed?  Would setting the dependencies down below one more time cause infinite loops?

Reply:
this function is called whenever the value of variable changes, could be two reason
* Manual user change (This can still happen during loading process as we do not block using the select during loading) 
* The value currently being updated has caused an update due to line 81